<h1><center>Laboratorio 6: La solicitud de Sergio 🤗</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2024</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Eduardo Moya
- Ayudantes: Nicolás Ojeda, Melanie Peña, Valentina Rojas

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Elizabeth Ramírez Z.
- Nombre de alumno 2: Lucas Orellana J.


### **Link de repositorio de GitHub:** [Insertar Repositorio](https://github.com/ElLuquitas/MDS7202-labs)

## Temas a tratar
- Aplicar Pandas para obtener características de un DataFrame.
- Aplicar Pipelines y Column Transformers.
- Utilizar diferentes algoritmos de cluster y ver el desempeño.

## Reglas:

- **Grupos de 2 personas**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias.
- Pueden usar cualquer matrial del curso que estimen conveniente.
- Código que no se pueda ejecutar, no será revisado.

### Objetivos principales del laboratorio
- Comprender cómo aplicar pipelines de Scikit-Learn para generar clusters.
- Familiarizarse con plotly.

El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `numpy`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre arreglos (*o tensores*).

## Descripción del laboratorio

<center>
<img src="https://i.pinimg.com/originals/5a/a6/af/5aa6afde8490da403a21601adf7a7240.gif" width=400 />

En el corazón de las operaciones de Aerolínea Lucero, Sergio, el gerente de análisis de datos, reunió a un talentoso equipo de jóvenes científicos de datos para un desafío crucial: segmentar la base de datos de los clientes. “Nuestro objetivo es descubrir patrones en el comportamiento de los pasajeros que nos permitan personalizar servicios y optimizar nuestras campañas de marketing,” explicó Sergio, mientras desplegaba un amplio rango de datos que incluían desde hábitos de compra hasta opiniones sobre los vuelos.

Sergio encargó a los científicos de datos la tarea de aplicar técnicas avanzadas de clustering para identificar distintos segmentos de clientes, como los viajeros frecuentes y aquellos que eligen la aerolínea para celebrar ocasiones especiales. La meta principal era entender profundamente cómo estos grupos perciben la calidad y satisfacción de los servicios ofrecidos por la aerolínea.

A través de un enfoque meticuloso y colaborativo, los científicos de datos se abocaron a la tarea, buscando transformar los datos brutos en valiosos insights que permitirían a Aerolínea Lucero no solo mejorar su servicio, sino también fortalecer las relaciones con sus clientes mediante una oferta más personalizada y efectiva.

## Importamos librerias utiles 😸

In [69]:
import numpy as np
import pandas as pd

from sklearn import datasets

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.mixture import GaussianMixture
from sklearn.pipeline import Pipeline
import matplotlib.pyplot as plt


## 1. Estudio de Performance 📈 [10 Puntos]



<center>
<img src="https://user-images.githubusercontent.com/57133330/188281408-c67df9ee-fd1f-4b37-833b-f02848f1ce02.gif" width=300>

Don Sergio les ha encomendado su primera tarea: analizar diversas técnicas de clustering. Su objetivo es entender detalladamente cómo funcionan estos métodos en términos de segmentación y eficiencia en tiempo de ejecución.

Analice y compare el desempeño, tiempo de ejecución y visualizaciones de cuatro algoritmos de clustering (k-means, DBSCAN, Ward y GMM) aplicados a tres conjuntos de datos, incrementando progresivamente su tamaño. Utilice Plotly para las gráficas y discuta los resultados tanto cualitativa como cuantitativamente.

Uno de los requisitos establecidos por Sergio es que el análisis se lleve a cabo utilizando Plotly; de no ser así, se considerará incorrecto. Para facilitar este proceso, se ha proporcionado un código de Plotly que puede servir como base para realizar las gráficas. Apóyese en el código entregado para efectuar el análisis y tome como referencia la siguiente imagen para realizar los gráficos:

<img src='https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/misc_images/Screenshot_2024-04-26_at_9.10.44_AM.png' width=800 />

En el gráfico se visualizan en dos dimensiones los diferentes tipos de datos proporcionados en `datasets`. Cada columna corresponde a un modelo de clustering diferente, mientras que cada fila representa un conjunto de datos distinto. Cada uno de los gráficos incluye el **tiempo en segundos que tarda el análisis** y la **métrica Silhouette obtenida**.

Para ser más específicos, usted debe cumplir los siguientes objetivos:
1. Generar una función que permita replicar el gráfico expuesto en la imagen (no importa que los colores calcen). [4 puntos]
2. Ejecuta la función para un `n_samples` igual a 1000, 5000, 10000. [2 puntos]
3. Analice y compare el desempeño, tiempo de ejecución y visualizaciones de cuatro algoritmos de clustering utilizando las 3 configuraciones dadas en `n_samples`. [4 puntos]


> ❗ Tiene libertad absoluta de escoger los hiper parámetros de los cluster, sin embargo, se recomienda verificar el dominio de las variables para realizar la segmentación.

> ❗ Recuerde que es obligatorio el uso de plotly.


In [70]:
"""
En la siguiente celda se crean los datos ficticios a usar en la sección 1 del lab.
❗No realice cambios a esta celda a excepción de n_samples❗
"""

# Datos a utilizar

# Configuracion
n_samples = 5000 #Este parámetro si lo pueden modificar

def create_data(n_samples):

    # Lunas
    moons = datasets.make_moons(n_samples=n_samples, noise=0.05, random_state=30)
    # Blobs
    blobs = datasets.make_blobs(n_samples=n_samples, random_state=172)
    # Datos desiguales
    transformation = [[0.6, -0.6], [-0.4, 0.8]]
    mutated = (np.dot(blobs[0], transformation), blobs[1])

    # Generamos Dataset
    dataset = {
        'moons':{
            'x': moons[0], 'classes': moons[1], 'n_cluster': 2
        },
        'blobs':{
            'x': blobs[0], 'classes': blobs[1], 'n_cluster': 3
        },
        'mutated':{
            'x': mutated[0], 'classes': mutated[1], 'n_cluster': 3
        }
    }
    return dataset

data_sets = create_data(n_samples)

**Respuestas:**

In [71]:
def plot_scatter(x, y, color, exe_time, silhouette_score):
    '''
    Entrega de manera gráfica un scatter plot de los datos entregados.

    Args:
        x (np.array): Coordenadas en x de los puntos a graficar (observaciones).
        y (np.array): Coordenadas en y de los puntos a graficar (clases).
        color (np.array): Clases a graficar diferenciadas por colores.
        exe_time (float): Tiempo de ejecución del algoritmo.
        silhouette_score (float): Valor del silhouette score.

    Returns:
        fig (plotly.graph_objects.Figure): Gráfico de puntos.
    '''
    fig = go.Scatter(
        x = x,
        y = y,
        mode = 'markers',
        marker = dict(
            color = color,
            colorscale = 'Viridis',
            showscale = False
        ),
        text = f"{exe_time:.2f} [s] | s: {silhouette_score:.2f}"
    )
    return fig

def plot_multi_scatter(matrix_data_sets):
    '''
    Entrega de manera gráfica un scatter plot de los datos ya clusterizados y
    que vienen dentro de una matriz.

    Args:
        matrix_data_sets (list): Lista con los datos a graficar, ordenados como matriz.
    '''
    fig = make_subplots(
        rows = 3,
        cols = 4,
        subplot_titles = ['Kmeans', 'GMM', 'WARD', 'DBSCAN']
    )

    # Iteración sobre los datasets
    for i, (data, exe_time, silhouette_score) in enumerate(matrix_data_sets):
        fig.add_trace(
            plot_scatter(data['x'], data['y'], data['pred'], exe_time, silhouette_score),
            row = i // 4 + 1, col = i % 4 + 1
        )

    fig.update_layout(height=800, width=1000, title_text="Comparación de métodos de clustering", showlegend=False)
    
    fig.show()

In [64]:
def plot_scatter(x, y, color):
    '''
    Entrega de manera gráfica un scatter plot de los datos entregados.

    Args:
        x (np.array): Coordenadas en x de los puntos a graficar (observaciones).
        y (np.array): Coordenadas en y de los puntos a graficar (clases).
        color (np.array): Clases a graficar diferenciadas por colores.

    Returns:
        fig (plotly.graph_objects.Figure): Gráfico de puntos.
    '''
    fig = go.Scatter(
        x = x,
        y = y,
        mode = 'markers',
        marker = dict(
            color = color,
            colorscale = 'Viridis',
            showscale = False
        )
    )
    return fig

def plot_multi_scatter(matrix_data_sets):
    '''
    Entrega de manera gráfica un scatter plot de los datos ya clusterizados y
    que vienen dentro de una matriz.

    Args:
        matrix_data_sets (list): Lista con los datos a graficar, ordenados como matriz.
    '''
    fig = make_subplots(
        rows = 3,
        cols = 4,
        subplot_titles = ['Kmeans', 'GMM', 'WARD', 'DBSCAN']
    )

    # Iteración sobre los datasets
    for i, (data, exe_time, silhouette_score) in enumerate(matrix_data_sets):
        fig.add_trace(
            plot_scatter(data['x'], data['y'], data['pred']),
            row = i // 4 + 1, col = i % 4 + 1
        )

        # Agregar el tiempo y el silhouette en la esquina superior derecha
        fig.add_annotation(
            text=f"{exe_time:.2f} [s] | s: {silhouette_score:.2f}",
            xref="x domain",
            yref="y domain",
            x=1, y=1,
            xanchor="right",
            yanchor="top",
            showarrow=False,
            font=dict(size=10),
            row=i // 4 + 1,
            col=i % 4 + 1

            
        )

    fig.update_layout(height=800, width=1000, title_text="Comparación de métodos de clustering", showlegend=False)
    
    fig.show()

In [72]:
#Se crean los tres dataset
data_sets_1 = create_data(1000)
data_sets_2 = create_data(5000)
data_sets_3 = create_data(10000)

In [73]:
# Aplicar algoritmos de clustering sobre los datos ficticios
from sklearn.cluster import KMeans, DBSCAN, AgglomerativeClustering
from sklearn.mixture import GaussianMixture 
from sklearn.metrics import silhouette_score
import time

# Función que se encarga del proceso de clustering
def cluster_model_apply(data_set, model, random_state = 30):
    '''
    Aplica un algoritmo de clustering sobre un dataset y devuelve el resultado
    en un DataFrame, el tiempo de ejecución y el silhouette score.

    Args:
        data_set (dict): Diccionario con los datos a clusterizar.
        model (sklearn.base.BaseEstimator): Algoritmo de clustering a aplicar.

    Returns:
        df (pd.DataFrame): DataFrame con las predicciones del modelo.
        exe_time (float): Tiempo de ejecución del algoritmo.
        silhouette_score (float): Valor del silhouette score.
    '''
    # Extraer datos

    x = data_set['x']
    classes = data_set['classes']
    n_cluster = data_set['n_cluster']

    # Aplicación de modelo y configuración dependiendo de cuál sea
    model_instance = model()
    if model == KMeans:
        model_instance = model_instance.set_params(n_clusters = n_cluster, random_state = random_state)

    elif model == GaussianMixture:
        model_instance = model_instance.set_params(n_components = n_cluster, random_state = random_state)

    elif model == AgglomerativeClustering:
        model_instance = model_instance.set_params(n_clusters = n_cluster, linkage = 'ward')

    elif model == DBSCAN:
        model_instance = model_instance.set_params(eps = 0.4, min_samples = 3)
    
    init = time.time()
    pred = model_instance.fit_predict(x)
    end = time.time()

    # Tiempo de ejecución
    exe_time = (end - init).__round__(2)

    # Calcular silhouette score (excepción creada para DBSCAN que fallaba)
    if len(set(pred)) > 1:
        silhouette = silhouette_score(x, pred).__round__(2)
    else:
        silhouette = float('nan')

    # Crear DataFrame
    df = pd.DataFrame({
        'x': x[:, 0],
        'y': x[:, 1],
        'class': classes,
        'pred': pred
    })

    return df, exe_time, silhouette

In [74]:
# Aplicación de Kmeans
data_sets_1_kmeans = [cluster_model_apply(data_sets_1[key], KMeans) for key in data_sets_1]
data_sets_2_kmeans = [cluster_model_apply(data_sets_2[key], KMeans) for key in data_sets_2]
data_sets_3_kmeans = [cluster_model_apply(data_sets_3[key], KMeans) for key in data_sets_3]

# Aplicación de GMM
data_sets_1_gmm = [cluster_model_apply(data_sets_1[key], GaussianMixture) for key in data_sets_1]
data_sets_2_gmm = [cluster_model_apply(data_sets_2[key], GaussianMixture) for key in data_sets_2]
data_sets_3_gmm = [cluster_model_apply(data_sets_3[key], GaussianMixture) for key in data_sets_3]

# Aplicación de Ward
data_sets_1_ward = [cluster_model_apply(data_sets_1[key], AgglomerativeClustering) for key in data_sets_1]
data_sets_2_ward = [cluster_model_apply(data_sets_2[key], AgglomerativeClustering) for key in data_sets_2]
data_sets_3_ward = [cluster_model_apply(data_sets_3[key], AgglomerativeClustering) for key in data_sets_3]

# Aplicación de DBSCAN
data_sets_1_dbscan = [cluster_model_apply(data_sets_1[key], DBSCAN) for key in data_sets_1]
data_sets_2_dbscan = [cluster_model_apply(data_sets_2[key], DBSCAN) for key in data_sets_2]
data_sets_3_dbscan = [cluster_model_apply(data_sets_3[key], DBSCAN) for key in data_sets_3]


In [75]:
data_sets_1000 = [data_sets_1_kmeans[0], data_sets_1_gmm[0], data_sets_1_ward[0], data_sets_1_dbscan[0],
                  data_sets_1_kmeans[1], data_sets_1_gmm[1], data_sets_1_ward[1], data_sets_1_dbscan[1],
                  data_sets_1_kmeans[2], data_sets_1_gmm[2], data_sets_1_ward[2], data_sets_1_dbscan[2]]

data_sets_5000 = [data_sets_2_kmeans[0], data_sets_2_gmm[0], data_sets_2_ward[0], data_sets_2_dbscan[0],
                  data_sets_2_kmeans[1], data_sets_2_gmm[1], data_sets_2_ward[1], data_sets_2_dbscan[1],
                  data_sets_2_kmeans[2], data_sets_2_gmm[2], data_sets_2_ward[2], data_sets_2_dbscan[2]]

data_sets_10000 = [data_sets_3_kmeans[0], data_sets_3_gmm[0], data_sets_3_ward[0], data_sets_3_dbscan[0],
                  data_sets_3_kmeans[1], data_sets_3_gmm[1], data_sets_3_ward[1], data_sets_3_dbscan[1],
                  data_sets_3_kmeans[2], data_sets_3_gmm[2], data_sets_3_ward[2], data_sets_3_dbscan[2]]

**1. Data sets 1000**

In [ ]:
plot_multi_scatter(data_sets_1000)

**2. Data sets 5000**

In [ ]:
plot_multi_scatter(data_sets_5000)

**3. Data sets 10000**

In [ ]:
plot_multi_scatter(data_sets_10000)

**3.Analice y compare el desempeño, tiempo de ejecución y visualizaciones de cuatro algoritmos de clustering utilizando las 3 configuraciones dadas en n_samples.**

* **Datasets con 1.000 datos**
    * En el primer y segundo grupo de datos ficticios (2 primeras filas de plot) Kmeans, GMM y Ward tienen cluster similares. El único diferente es DBSCAN.
    * En el tercer conjunto de datos (tercera fila) hay más diferencias entre todos los métodos de clustering. 

* **Datasets con 5.000 datos**
    * En el primer y segundo grupo de datos tienústeruste casi iales en Kmeans, GMM . Difiere un poco en el primer grupo de datos Ward. De nuevo DBSCAN es el más diferente. 
    * En el segundo grupo de datos, Ward difiere de Kmeans y GMM. Nueva,mente DBSCAN es el más disímil. 
    * En el tercer grupo de datos (tercera fila) se ve que Kmeans, GMM y Ward tiene un cluster que es igual. En los otros 2 cluster difieren. Nuevamente DBSCAN es el más diferente.  

* **Datasets con 10.000 datos**
    * Kmeans y GMM es prácticamente igual en las dos primeras filas de grupos de datos. Ward difiere de Kmeand y GMM en los colores de sus cluste en el primer y segundo grupo. 
    * Nuevmanete DBSCAN es el más distintivo. 

En términos generales se ve que Kmeans y GMM dan resultados bastante similares, a pesar de ir variando el número de elemntos en el dataset. Ward se asemeja a Kmeans y GMM dependiendo de la distribución de los datos. DBSCAN es el método que entrega resultados más diferentes en todos los casos. Para todos los dataset, GMM fue el que más tardó, mientras que Kmeans el que tendió a tardar menos.  

El índice Silhouette evalúa la calidad del agrupamiento obtenido por algoritmo. Su valor ideal es 1, ya que significa que los clusters Están bien definidos entre sí. Un valor 0 o menor indica que el agrupamiento no está siendo adecuado, es decir, no se están logrando "agrupar" y que se logre el objetivo de la clusterización.

En este caso, se observa que el valor de índice depende de los datos y la técnica aplicada. Por ejemplo, en la primera fila de datos, se logra mayor índice con Kmeans y los valores más bajo con DBSCAN. En el grupo de la segunda fila, se logran índices mucho más altos, donde Kmeans, GMM y Ward obtienen valores altos (índice mayor a 0.75) independiente del número de datos en el dataset. En la tercera fila de datos, se logran valores cercano a 0.5 con todos los métodos.

## 2. Análisis de Satisfacción de Vuelos. [10 puntos]

<center>
<img src="https://i.gifer.com/2Hci.gif" width=400 />

Habiendo entendido cómo funcionan los modelos de aprendizaje no supervisado, *Don Sergio* le encomienda estudiar la satisfacción de pasajeros al haber tomado un vuelo en alguna de sus aerolineas. Para esto, el magnate le dispone del dataset `aerolineas_licer.parquet`, el cual contiene el grado de satisfacción de los clientes frente a diferentes aspectos del vuelo. Las características del vuelo se definen a continuación:

- *Gender*: Género de los pasajeros (Femenino, Masculino)
- *Customer Type*: Tipo de cliente (Cliente habitual, cliente no habitual)
- *Age*: Edad actual de los pasajeros
- *Type of Travel*: Propósito del vuelo de los pasajeros (Viaje personal, Viaje de negocios)
- *Class*: Clase de viaje en el avión de los pasajeros (Business, Eco, Eco Plus)
- *Flight distance*: Distancia del vuelo de este viaje
- *Inflight wifi service*: Nivel de satisfacción del servicio de wifi durante el vuelo (0:No Aplicable; 1-5)
- *Departure/Arrival time convenient*: Nivel de satisfacción con la conveniencia del horario de salida/llegada
- *Ease of Online booking*: Nivel de satisfacción con la facilidad de reserva en línea
- *Gate location*: Nivel de satisfacción con la ubicación de la puerta
- *Food and drink*: Nivel de satisfacción con la comida y la bebida
- *Online boarding*: Nivel de satisfacción con el embarque en línea
- *Seat comfort*: Nivel de satisfacción con la comodidad del asiento
- *Inflight entertainment*: Nivel de satisfacción con el entretenimiento durante el vuelo
- *On-board service*: Nivel de satisfacción con el servicio a bordo
- *Leg room service*: Nivel de satisfacción con el espacio para las piernas
- *Baggage handling*: Nivel de satisfacción con el manejo del equipaje
- *Check-in service*: Nivel de satisfacción con el servicio de check-in
- *Inflight service*: Nivel de satisfacción con el servicio durante el vuelo
- *Cleanliness*: Nivel de satisfacción con la limpieza
- *Departure Delay in Minutes*: Minutos de retraso en la salida
- *Arrival Delay in Minutes*: Minutos de retraso en la llegada

En consideración de lo anterior, realice las siguientes tareas:

0. Ingeste el dataset a su ambiente de trabajo.

1. Seleccione **sólo las variables numéricas del dataset**.  Explique qué éfectos podría causar el uso de variables categóricas en un algoritmo no supervisado. [2 punto]

2. Realice una visualización de la distribución de cada variable y analice cada una de estas distribuciones. [2 punto]

3. Basándose en los gráficos, evalúe la necesidad de escalar los datos y explique el motivo de su decisión. [2 puntos]

4. Examine la correlación entre las variables mediante un correlograma. [2 puntos]

5. De acuerdo con los resultados obtenidos en 5, reduzca la dimensionalidad del conjunto de datos a cuatro variables, justificando su elección respecto a las variables que decide eliminar. [2 puntos]

**Respuesta:**

In [93]:
# Carga de datos
df_aerolineas = pd.read_parquet('aerolineas_lucer.parquet')

# Variables numéricas del dataset
df_aerolineas_num = df_aerolineas.select_dtypes(include = ["number"])

# Visualización del nombre de las columnas
print('Columnas numéricas del dataset:')
display(df_aerolineas_num.columns)

# Visualización de las primeras filas del dataset
df_aerolineas_num.head()

Columnas numéricas del dataset:


Index(['id', 'Age', 'Flight Distance', 'Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Gate location', 'Food and drink', 'Online boarding', 'Seat comfort',
       'Inflight entertainment', 'On-board service', 'Leg room service',
       'Baggage handling', 'Checkin service', 'Inflight service',
       'Cleanliness', 'Departure Delay in Minutes',
       'Arrival Delay in Minutes'],
      dtype='object')

id  Age  Flight Distance  Inflight wifi service  \
0   70172   13              460                      3   
1    5047   25              235                      3   
2  110028   26             1142                      2   
3   24026   25              562                      2   
4  119299   61              214                      3   

   Departure/Arrival time convenient  Ease of Online booking  Gate location  \
0                                  4                       3              1   
1                                  2                       3              3   
2                                  2                       2              2   
3                                  5                       5              5   
4                                  3                       3              3   

   Food and drink  Online boarding  Seat comfort  Inflight entertainment  \
0               5                3             5                       5   
1               1                3             1                       1   
2               5                5             5                       5   
3               2                2             2                       2   
4               4                5             5                       3   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 4               3.0                 4              4.0   
1                 1               5.0                 3              1.0   
2                 4               3.0                 4              4.0   
3                 2               5.0                 3              1.0   
4                 3               4.0                 4              3.0   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0               5.0            5                          25   
1               4.0            1                           1   
2               4.0            5                           0   
3               4.0            2                          11   
4               3.0            3                           0   

   Arrival Delay in Minutes  
0                      18.0  
1                       6.0  
2                       0.0  
3                       9.0  
4                       0.0

In [94]:
# Dimensiones del dataset creado
print(f'Cantidad de observaciones: {df_aerolineas_num.shape[0]}')

# Contar valores únicos en el dataset
print(f'Valores únicos en el dataset:')
display(df_aerolineas_num.nunique())

# Contar valores nulos del dataset
print(f'Valores nulos en el dataset:')
display(df_aerolineas_num.isnull().sum())

Cantidad de observaciones: 236114
Valores únicos en el dataset:


id                                   103592
Age                                      75
Flight Distance                        3802
Inflight wifi service                     6
Departure/Arrival time convenient         6
Ease of Online booking                    6
Gate location                             6
Food and drink                            6
Online boarding                           6
Seat comfort                              6
Inflight entertainment                    6
On-board service                          6
Leg room service                          6
Baggage handling                          5
Checkin service                           6
Inflight service                          6
Cleanliness                               6
Departure Delay in Minutes              445
Arrival Delay in Minutes                455
dtype: int64

Valores nulos en el dataset:


id                                   0
Age                                  0
Flight Distance                      0
Inflight wifi service                0
Departure/Arrival time convenient    0
Ease of Online booking               0
Gate location                        0
Food and drink                       0
Online boarding                      0
Seat comfort                         0
Inflight entertainment               0
On-board service                     0
Leg room service                     0
Baggage handling                     0
Checkin service                      0
Inflight service                     0
Cleanliness                          0
Departure Delay in Minutes           0
Arrival Delay in Minutes             0
dtype: int64

In [95]:
df_aerolineas_num.describe()

id            Age  Flight Distance  Inflight wifi service  \
count  236114.000000  236114.000000    236114.000000          236114.000000   
mean    64976.757820      39.385297      1190.646891               2.729199   
std     37475.512146      15.125768       999.005937               1.326889   
min         1.000000       7.000000        31.000000               0.000000   
25%     32573.000000      27.000000       414.000000               2.000000   
50%     64945.000000      40.000000       843.000000               3.000000   
75%     97414.000000      51.000000      1744.000000               4.000000   
max    129880.000000      85.000000      4983.000000               5.000000   

       Departure/Arrival time convenient  Ease of Online booking  \
count                      236114.000000           236114.000000   
mean                            3.060598                2.758464   
std                             1.524643                1.399083   
min                             0.000000                0.000000   
25%                             2.000000                2.000000   
50%                             3.000000                3.000000   
75%                             4.000000                4.000000   
max                             5.000000                5.000000   

       Gate location  Food and drink  Online boarding   Seat comfort  \
count  236114.000000   236114.000000    236114.000000  236114.000000   
mean        2.978536        3.202478         3.250218       3.439584   
std         1.278132        1.328406         1.349914       1.317642   
min         0.000000        0.000000         0.000000       0.000000   
25%         2.000000        2.000000         2.000000       2.000000   
50%         3.000000        3.000000         3.000000       4.000000   
75%         4.000000        4.000000         4.000000       5.000000   
max         5.000000        5.000000         5.000000       5.000000   

       Inflight entertainment  On-board service  Leg room service  \
count           236114.000000     236114.000000     236114.000000   
mean                 3.360131          3.383416          3.353338   
std                  1.332025          1.287594          1.315401   
min                  0.000000          0.000000          0.000000   
25%                  2.000000          2.000000          2.000000   
50%                  4.000000          4.000000          4.000000   
75%                  4.000000          4.000000          4.000000   
max                  5.000000          5.000000          5.000000   

       Baggage handling  Checkin service  Inflight service    Cleanliness  \
count     236114.000000    236114.000000     236114.000000  236114.000000   
mean           3.632991         3.303980          3.641978       3.286396   
std            1.180697         1.266034          1.174758       1.311961   
min            1.000000         0.000000          0.000000       0.000000   
25%            3.000000         3.000000          3.000000       2.000000   
50%            4.000000         3.000000          4.000000       3.000000   
75%            5.000000         4.000000          5.000000       4.000000   
max            5.000000         5.000000          5.000000       5.000000   

       Departure Delay in Minutes  Arrival Delay in Minutes  
count               236114.000000             236114.000000  
mean                    14.765512                 15.213676  
std                     38.114330                 38.705523  
min                      0.000000                  0.000000  
25%                      0.000000                  0.000000  
50%                      0.000000                  0.000000  
75%                     12.000000                 13.000000  
max                   1592.000000               1584.000000

**2. Explique qué efectos podría causar el uso de variables categóricas en un algoritmo no supervisado.**
El uso de variables categóricas sin tener un preprocesamiento en algoritmos no supervisados, como clustering, puede generar varios problemas:

1. **Cálculo de distancias no aplicable**: Los algoritmos basados en distancias, como K-Means, no pueden calcular distancias entre variables categóricas, lo que genera agrupaciones incorrectas cuando se usan variable categórica. 

2. **Escalado desbalanceado**: Cuando convertimos las variables categóricas en varias columnas usando One-Hot Encoding (esta es una técnica para transformar variables categóricas), éstas pueden tener más influencia de lo que deberían en los resultados del agrupamiento, lo que puede distorsionar el análisis.

3. **Agrupaciones no significativas**: Las categorías pueden ser interpretadas como números arbitrarios, lo que genera grupos sin sentido.

4. **Pérdida de información semántica**: Sin preprocesamiento adecuado, se pierde el significado importante de las categorías, afectando la calidad de los resultados.

### Visualizaciones de gráficos

In [ ]:
# Histograma para cada una de las variables presentes en 'df_aerolineas_num'
for col in df_aerolineas_num.columns:
    fig = px.histogram(df_aerolineas_num, marginal="box", x=col, title=f'Histograma de {col}')
    fig.show()

**3. Analice cada una de estas distribuciones**

Viendo cómo están distribuidas cada una de estas características, podemos notar que:
* `id`: Dentro del dataset existen, para algunos clientes, más de una observación, lo cual indica que han tomado más de un vuelo y añadiendo así más información para un mismo cliente, teniendo un total de $103.592$ clientes en las $236.114$ observaciones totales. La distribución del ID es una distribución uniforme, es decir, que todos los ID tienen similares frecuencias. 
* `Age`: Se tiene un rango amplio de clientes, estando el grueso centrado entre los 20 y 70 años. Se observa una distribución con dos peak, pareciera ser una distribución bimodal, es decir, con dos modas en sus datos. Un peak es en torno a los años 25 y el otro peak es en torno a los 39 años.  
* `Flight Distance`: Las distancias son en su mayoría en recorridos cortos, habiendo una concentración de distancias entre las $60$ y $1.200$ unidades de distancia. La distribución de los datos es asimétrica. En este caso, se ve que hay una asimetría hacia la derecha. También hay una cola hacia la izquierda de los datos, entre los 1300 a 4000 km (supondremos que la distancia es en km, aunque podría ser en otra unidad de medida como millas). 
* `Inflight wifi service`: Calificación con moda entre $2$ y $3$. Al parecer no es muy bueno. Esta distribución se asemejan a una normal, con una moda, que pareciera coincidir con la mediana y media. La única salvedad para que sea una distribución normal del todo, es que la cola derecha tiene en la última parte más frecuencia de datos que la última parte de la cola de la izquierda. 
* `Departure/Arrival time convenient`: Calificación con moda en $4$. Esta distribución tiene asimetría negativa o hacia la izquierda, ya que los datos se concentran hacia los mayores valores. 
* `Ease of Online booking`: Calificación con moda en $3$. Tiene datos que se asemejan a una distribución normal. Esta distribución de datos se asemeja a lo visto en la distribución Inflight wifi service.
* `Gate location`: Calificación con moda en $3$
* `Food and drink`: Calificación con moda en $4$
* `Online boarding`: Calificación con moda en $4$
* `Seat comfort`: Calificación con moda en $4$
* `Inflight entertainment`: Calificación con moda en $4$
* `On-board service`: Calificación con moda en $4$
* `Leg room service`: Calificación con moda en $4$
* `Baggage handling`: Calificación con moda en $4$
* `Checkin service`: Calificación con moda en $4$ y bastante pareja con $3$. Los datos se asemejan a una distribución normal. 
* `Inflight service`: Calificación con moda en $4$. Los datos se concentran hacia la derecha, por lo que la distribución tiene asimetría. 
* `Cleanliness`: Calificación con moda en $4$. Tiene una distribución que concentra los datos hacia la derecha. 
* `Departure Delay in Minutes`: Se tiene una gran concentración de observaciones que no presentan retraso en la salida del vuelo, representando un poco más de la mitad de las observaciones.Tiene una distribución logarítmica o exponencial
* `Arrival Delay in Minutes`: Lo mismo que el caso anterior. La gran mayoría de vuelos no presenta retrasos en el arribo a destino. Tiene una distribución logarítmica o exponencial. 

**3. Basándose en los gráficos, evalúe la necesidad de escalar los datos y explique el motivo de su decisión.**

Existe una necesidad de escalar los datos (a excepción de `id`) especialmente por las variables que representan tiempo y distancia, ya que tienen rangos muy amplios de valores, concentrándose la mayoría de ellos en un extremo del eje `x`. Si no se tienen los datos escalados, la comparación entre variables se hace muy dificultosa, pudiendo generar sesgos al implementar algoritmos de clusterización. De manera más ordenada, se puede decir que:

1. Rango de variación de las variables: al ver los histogramas, se observa que los datos no varían en los mismos rangos. Por ello sería mejor que todos fluctúen entre 0 y 1, para que tengan el mismo peso/influencia en el modelo. 
2. Distribución: hay algunas variables que tienen concentrados sus montos en valores altos o bajos. Al escalar los datos, se homogeneiza el rango de variación de las variables y de alguna manera compensa esto. 
3. Algunas técnicas de clustering se basan en distancias, por lo que es importante que las variables que tienen rangos de valores mayores no pesen más que las variables que fluctúan en rangos más acostados. 

Por ejemplo, la variable `Arrival Delay in Minutes` varía de valores cercanos a 0 hasta sobre 1400 minutos, mientras que la variable `Departure/Arrival time convenient` varía de 0 a 5. Los rangos de ambas variable numérica son bastante diferentes, por ello es esencial escalar los datos. 



### Escalamiento de datos

In [96]:
# Escalamiento de datos a través de la función MinMaxScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
import time

numerical_columns = df_aerolineas_num.drop(columns = ['id']).columns

column_transformer = ColumnTransformer(
    [('numerical', MinMaxScaler(), numerical_columns)],
    remainder='passthrough'
)

pipeline = Pipeline([('scaler', column_transformer)])

# Aplicamos ColumnTransformer a los datos
init = time.time()
df_transformed = pd.DataFrame(
    pipeline.fit_transform(df_aerolineas_num.drop(columns = ['id'])),
    columns = numerical_columns
)
end = time.time()
print('Tiempo de ejecución (en segundos):', end - init)

Tiempo de ejecución (en segundos): 0.48551511764526367


In [97]:
df_transformed.head()

Age  Flight Distance  Inflight wifi service  \
0  0.076923         0.086632                    0.6   
1  0.230769         0.041195                    0.6   
2  0.243590         0.224354                    0.4   
3  0.230769         0.107229                    0.4   
4  0.692308         0.036955                    0.6   

   Departure/Arrival time convenient  Ease of Online booking  Gate location  \
0                                0.8                     0.6            0.2   
1                                0.4                     0.6            0.6   
2                                0.4                     0.4            0.4   
3                                1.0                     1.0            1.0   
4                                0.6                     0.6            0.6   

   Food and drink  Online boarding  Seat comfort  Inflight entertainment  \
0             1.0              0.6           1.0                     1.0   
1             0.2              0.6           0.2                     0.2   
2             1.0              1.0           1.0                     1.0   
3             0.4              0.4           0.4                     0.4   
4             0.8              1.0           1.0                     0.6   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0               0.8               0.6              0.75              0.8   
1               0.2               1.0              0.50              0.2   
2               0.8               0.6              0.75              0.8   
3               0.4               1.0              0.50              0.2   
4               0.6               0.8              0.75              0.6   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0               1.0          1.0                    0.015704   
1               0.8          0.2                    0.000628   
2               0.8          1.0                    0.000000   
3               0.8          0.4                    0.006910   
4               0.6          0.6                    0.000000   

   Arrival Delay in Minutes  
0                  0.011364  
1                  0.003788  
2                  0.000000  
3                  0.005682  
4                  0.000000

In [98]:
df_transformed.describe()

Age  Flight Distance  Inflight wifi service  \
count  236114.000000    236114.000000          236114.000000   
mean        0.415196         0.234177               0.545840   
std         0.193920         0.201738               0.265378   
min         0.000000         0.000000               0.000000   
25%         0.256410         0.077342               0.400000   
50%         0.423077         0.163974               0.600000   
75%         0.564103         0.345921               0.800000   
max         1.000000         1.000000               1.000000   

       Departure/Arrival time convenient  Ease of Online booking  \
count                      236114.000000           236114.000000   
mean                            0.612120                0.551693   
std                             0.304929                0.279817   
min                             0.000000                0.000000   
25%                             0.400000                0.400000   
50%                             0.600000                0.600000   
75%                             0.800000                0.800000   
max                             1.000000                1.000000   

       Gate location  Food and drink  Online boarding   Seat comfort  \
count  236114.000000   236114.000000    236114.000000  236114.000000   
mean        0.595707        0.640496         0.650044       0.687917   
std         0.255626        0.265681         0.269983       0.263528   
min         0.000000        0.000000         0.000000       0.000000   
25%         0.400000        0.400000         0.400000       0.400000   
50%         0.600000        0.600000         0.600000       0.800000   
75%         0.800000        0.800000         0.800000       1.000000   
max         1.000000        1.000000         1.000000       1.000000   

       Inflight entertainment  On-board service  Leg room service  \
count           236114.000000     236114.000000     236114.000000   
mean                 0.672026          0.676683          0.670668   
std                  0.266405          0.257519          0.263080   
min                  0.000000          0.000000          0.000000   
25%                  0.400000          0.400000          0.400000   
50%                  0.800000          0.800000          0.800000   
75%                  0.800000          0.800000          0.800000   
max                  1.000000          1.000000          1.000000   

       Baggage handling  Checkin service  Inflight service    Cleanliness  \
count     236114.000000    236114.000000     236114.000000  236114.000000   
mean           0.658248         0.660796          0.728396       0.657279   
std            0.295174         0.253207          0.234952       0.262392   
min            0.000000         0.000000          0.000000       0.000000   
25%            0.500000         0.600000          0.600000       0.400000   
50%            0.750000         0.600000          0.800000       0.600000   
75%            1.000000         0.800000          1.000000       0.800000   
max            1.000000         1.000000          1.000000       1.000000   

       Departure Delay in Minutes  Arrival Delay in Minutes  
count               236114.000000             236114.000000  
mean                     0.009275                  0.009605  
std                      0.023941                  0.024435  
min                      0.000000                  0.000000  
25%                      0.000000                  0.000000  
50%                      0.000000                  0.000000  
75%                      0.007538                  0.008207  
max                      1.000000                  1.000000

### Visualización de correlación

In [ ]:
correlations = df_transformed.corr()
px.imshow(correlations,
          labels=dict(x="Variables", y="Variables", color="Correlación"),
          title="Mapa de calor de correlaciones",
          width=800, height=800)

In [100]:
correlations.head()

Age  Flight Distance  \
Age                                1.000000         0.101150   
Flight Distance                    0.101150         1.000000   
Inflight wifi service              0.017572         0.007187   
Departure/Arrival time convenient  0.039008        -0.019793   
Ease of Online booking             0.024927         0.066585   

                                   Inflight wifi service  \
Age                                             0.017572   
Flight Distance                                 0.007187   
Inflight wifi service                           1.000000   
Departure/Arrival time convenient               0.343414   
Ease of Online booking                          0.715727   

                                   Departure/Arrival time convenient  \
Age                                                         0.039008   
Flight Distance                                            -0.019793   
Inflight wifi service                                       0.343414   
Departure/Arrival time convenient                           1.000000   
Ease of Online booking                                      0.437551   

                                   Ease of Online booking  Gate location  \
Age                                              0.024927      -0.001224   
Flight Distance                                  0.066585       0.004086   
Inflight wifi service                            0.715727       0.335391   
Departure/Arrival time convenient                0.437551       0.445463   
Ease of Online booking                           1.000000       0.459608   

                                   Food and drink  Online boarding  \
Age                                      0.022470         0.207535   
Flight Distance                          0.057216         0.216355   
Inflight wifi service                    0.134076         0.456805   
Departure/Arrival time convenient        0.004046         0.069607   
Ease of Online booking                   0.030885         0.403554   

                                   Seat comfort  Inflight entertainment  \
Age                                    0.158939                0.076808   
Flight Distance                        0.158620                0.129742   
Inflight wifi service                  0.122673                0.209171   
Departure/Arrival time convenient      0.010475               -0.005432   
Ease of Online booking                 0.029723                0.045944   

                                   On-board service  Leg room service  \
Age                                        0.059467          0.042472   
Flight Distance                            0.109537          0.134848   
Inflight wifi service                      0.123733          0.160075   
Departure/Arrival time convenient          0.068800          0.011975   
Ease of Online booking                     0.039992          0.105803   

                                   Baggage handling  Checkin service  \
Age                                       -0.046668         0.034976   
Flight Distance                            0.064242         0.073001   
Inflight wifi service                      0.121764         0.043869   
Departure/Arrival time convenient          0.071662         0.092419   
Ease of Online booking                     0.038811         0.010731   

                                   Inflight service  Cleanliness  \
Age                                       -0.048948     0.052684   
Flight Distance                            0.057966     0.093521   
Inflight wifi service                      0.111062     0.131513   
Departure/Arrival time convenient          0.072624     0.013094   
Ease of Online booking                     0.034344     0.014853   

                                   Departure Delay in Minutes  \
Age                                                 -0.009492   
Flight Distance                                      0.001517   
Inflight wifi service                               -0.0

### Reducción de dimensionalidad

Como se puede ver en el mapa de correlación, existen algunas características que están bastante relacionadas entre sí, como lo son:

* Arrival Delay in Minutes / Departure Delay in Minutes: 0.97
* Ease of Online Booking / Inflight wifi service: 0.72
* Inflight entertainment / Cleanliness: 0.69
* Seat comfort / Cleanliness: 0.68
* Food and drink / Cleanliness: 0.66
* Baggage adding / Inflight service: 0.63
* Food and drink / Inflight entertainment: 0.62
* Seat comfort / Iflight entertainment: 0.61

Si no queremos información redundante, lo recomendado sería eliminar algunas de estas características con alta correlación. Además, del resto de columnas que no presentan altas correlaciones, hay algunas que podemos eliminar dado que la información que aportan no es tan relevante. Con estas consideraciones previas, podemos quedarnos con las siguientes columnas:
* Departure Delay in Minutes
* Inflight wifi service
* Cleanliness
* Inflight entertainment




In [ ]:
# Correlación entre las variables seleccionadas
# Calcular la matriz de correlación
corr_matrix = df_transformed[['Inflight wifi service', 'Inflight entertainment', 'Cleanliness', 'Departure Delay in Minutes']].corr()

# Crear el gráfico de la matriz de correlación
fig = go.Figure(
    data=go.Heatmap(
        z=corr_matrix.values,
        x=corr_matrix.columns,
        y=corr_matrix.columns,
        colorscale='RdBu',  # Colores para mostrar correlación positiva/negativa
        zmin=-1, zmax=1,    # Rango de correlación entre -1 y 1
        showscale=True
    )
)

# Ajustar el layout del gráfico
fig.update_layout(
    title="Correlation Matrix",
    xaxis_title="Variables",
    yaxis_title="Variables",
    width=600, height=600
)

# Mostrar el gráfico
fig.show()

## 3. Preprocesamiento 🎭. [10 puntos]

<center>
<img src="https://i.pinimg.com/originals/1e/a8/0e/1ea80e7cea0d429146580c7e91c5b944.gif" width=400>

Tras quedar satisfecho con los resultados presentados en el punto 2, el dueño de la empresa ha solicitado que se preprocesen los datos mediante un `pipeline`. Es crucial que este proceso tenga en cuenta las observaciones derivadas de los análisis anteriores. Adicionalmente, ha expresado su interés en visualizar el conjunto de datos en un gráfico de dos o tres dimensiones.

Basándose en los análisis realizados anteriormente:
1. Cree un `pipeline` que incluya PCA, utilizando las consideraciones mencionadas previamente para proyectar los datos a dos dimensiones. [4 puntos]
2. Grafique los resultados obtenidos y comente lo visualizado. [6 puntos]

**Respuestas:**

Si consideramos realizar todos los pasos anteriores y colocarlos en un solo *pipeline*, el orden es el siguiente:

In [151]:
from sklearn.decomposition import PCA

# 1. Selección de variables numéricas del dataset original
numerical_columns = df_aerolineas_num.drop(columns = ['id']).columns
interest_columns = ['Inflight wifi service', 'Inflight entertainment', 'Cleanliness', 'Departure Delay in Minutes']#intento 1
#interest_columns = ['Gate location', 'Arrival Delay in Minutes', 'Departure/Arrival time convenient', 'Online boarding']
df_filtered = df_aerolineas_num.loc[:, interest_columns]

# 2. Instancia de MinMaxScaler y ColumnTransformer
column_transformer = ColumnTransformer(
    [('numerical', MinMaxScaler(), interest_columns)],
    remainder='passthrough'
)

# 3. Instancia de PCA
pca = PCA(n_components = 2)

# 4. Pipeline
pipeline = Pipeline([
    ('scaler', column_transformer),
    ('pca', pca)
])

# Aplicación y conversión a DataFrame
transform_data_pca = pipeline.fit_transform(df_filtered)
transform_data_pca = pd.DataFrame(transform_data_pca, columns=['PC1', 'PC2'])

In [152]:
# Un vistazo a los datos transformados
transform_data_pca.head()

PC1       PC2
0 -0.467886  0.094105
1  0.609520 -0.251048
2 -0.407051  0.283905
3  0.401034  0.025052
4  0.070803 -0.078477

In [ ]:
fig = px.scatter(transform_data_pca, x='PC1', y='PC2', title="Proyección PCA a 2 Dimensiones")
fig.show()

**2. Comente lo visualizado en el gráfico**

La gráfica muestra una proyección de PCA en dos dimensiones, donde los ejes representan los dos principales componentes (PC1 y PC2). Los puntos forman un patrón diagonal, lo que sugiere correlación entre estos componentes y que están capturando variaciones lineales de los datos originales. Se observan algunos agrupamientos, con 10 patrones diagonales. En resumen, los primeros dos componentes explican bastante variabilidad de los datos, pero la estructura lineal indica posibles correlaciones internas. Sé probo con otros sets de variables, pero no entregaban resultados razonables. 

Cabe señalar que el resultado del PCA depende del escalamiento que se realice, por ejemplo probamos con otros métodos de escalamiento, como RobustScaler que daba un resultado bastante diferente, mientras que un escalamiento con la media y desviación estándar daba un resultado similar. 

## 4. Outliers 🚫🙅‍♀️❌🙅‍♂️ [10 puntos]

<center>
<img src="https://joachim-gassen.github.io/images/ani_sim_bad_leverage.gif" width=250>

Con el objetivo de mantener la claridad en su análisis, Don Sergio le ha solicitado entrenar un modelo que identifique pasajeros con comportamientos altamente atípicos.

1. Utilice `IsolationForest` para clasificar las anomalías del dataset (sin aplicar PCA), configurando el modelo para que sólo el 1% de los datos sean considerados anómalos. Asegúrese de integrar esta tarea dentro de un `pipeline`. [3 puntos]

2. Visualice los resultados en el gráfico de dos dimensiones previamente creado. [3 puntos]

3. ¿Cómo evaluaría el rendimiento de su modelo en la detección de anomalías? [4 puntos]

**Respuestas:**

In [ ]:
from sklearn.ensemble import IsolationForest

# Instancia de IsolationForest
isf = IsolationForest(contamination = 0.01, random_state = 30)

# Creación de pipeline considerando puntos anteriores
pipe_isf = Pipeline([
    ('scaler', column_transformer),
    ('isf', isf)
])

# Aplicación de IsolationForest y conversión a DataFrame
transform_data_isf = pipe_isf.fit_predict(df_filtered)
outliers = pd.DataFrame(transform_data_isf, columns=['outlier'])
transform_data_pca['outlier'] = outliers

# Vemos los resultados en el PCA hecho anteriormente
fig = px.scatter(
    transform_data_pca,
    x='PC1',
    y='PC2',
    color='outlier',
    color_discrete_map={-1: 'red', 1: 'blue'},
    title="Proyección PCA a 2 Dimensiones con Isolation Forest para detección de outliers"
)

fig.show()

**3.¿Cómo evaluaría el rendimiento de su modelo en la detección de anomalías?**

Los valores de -1 son valores *outliers*. No es buena la detección, ya que los puntos no son outliers en la nube de puntos. Sin embargo, sabemos que depende mucho de los pasos previos que se hayan realizado. Por ejemplo, si se escalan los datos con el método `RobustScaler`, el resultado es distinto. Adjuntamos un resultado obtenido. Quizás, para detectar *outliers* con el escalamiento `MinMax`, es necesario aplicar otro método de detección de *outliers*. 

In [ ]:
# 1. Instancia de MinMaxScaler y ColumnTransformer
column_transformer = ColumnTransformer(
    [('numerical', RobustScaler(), interest_columns)],
    remainder='passthrough'
)

# 2. Instancia de PCA
pca = PCA(n_components = 2)

# 3. Pipeline
pipeline = Pipeline([
    ('scaler', column_transformer),
    ('pca', pca)
])

# Aplicación y conversión a DataFrame
transform_data_pca = pipeline.fit_transform(df_filtered)
transform_data_pca = pd.DataFrame(transform_data_pca, columns=['PC1', 'PC2'])

# Instancia de IsolationForest
isf = IsolationForest(contamination = 0.01, random_state = 30)

# Creación de pipeline considerando puntos anteriores
pipe_isf = Pipeline([
    ('scaler', column_transformer),
    ('isf', isf)
])

# Aplicación de IsolationForest y conversión a DataFrame
transform_data_isf = pipe_isf.fit_predict(df_filtered)
outliers = pd.DataFrame(transform_data_isf, columns=['outlier'])
transform_data_pca['outlier'] = outliers

# Vemos los resultados en el PCA hecho anteriormente
fig = px.scatter(
    transform_data_pca,
    x='PC1',
    y='PC2',
    color='outlier',
    color_discrete_map={-1: 'red', 1: 'blue'},
    title="Proyección PCA a 2 Dimensiones con Isolation Forest para detección de outliers"
)

fig.show()

**Análisis:**
Se observa, que al aplicar este método de escalamiento, los valores con -1 que serían los *outliers* tienen más sentido en los datos gráficos, ya que son esos puntos que se alejan de los datos que están acumulados hacia la izquierda. 

## 5. Métricas de Desempeño 🚀 [10 puntos]

<center>
<img src="https://giffiles.alphacoders.com/219/219081.gif" width=300>

Motivado por incrementar su fortuna, Don Sergio le solicita entrenar un modelo que le permita segmentar a los pasajeros en grupos distintos, con el objetivo de optimizar las diversas campañas de marketing diseñadas por su equipo. Para ello, le se pide realizar las siguientes tareas:

1. Utilizar el modelo **Gaussian Mixture** y explore diferentes configuraciones de número de clústers, específicamente entre 3 y 8. Asegúrese de integrar esta operación dentro de un `pipeline`. [4 puntos]
2. Explique cuál sería el criterio adecuado para seleccionar el número óptimo de clústers. **Justifique de forma estadistica y a traves de gráficos.** [6 puntos]

> **HINT:** Se recomienda investigar sobre los criterios AIC y BIC para esta tarea.

**Respuestas:**

In [156]:
def evaluate_gmm_aic_bic(df, variables, n_clusters_range):
    """
    Evalúa un modelo de mezcla gaussiana para diferentes números de clústers 
    y grafica las puntuaciones AIC y BIC.

    Parámetros:
        df (DataFrame): DataFrame que contiene los datos.
        variables (list): Lista de nombres de las columnas a usar.
        n_clusters_range (range): Rango de números de clústers a explorar.
    """
    # Definir el pipeline
    pipeline = Pipeline([
        ('gmm', GaussianMixture(random_state=42))
    ])

    # Listas para almacenar los resultados de AIC y BIC
    aic_scores = []
    bic_scores = []

    # Iterar sobre los diferentes números de clústers
    for n_clusters in n_clusters_range:
        # Ajustar el modelo con el número de clústers actual
        pipeline.set_params(gmm__n_components=n_clusters)
        pipeline.fit(df[variables])  # Usar solo las columnas seleccionadas
        
        # Obtener las métricas AIC y BIC
        gmm_model = pipeline.named_steps['gmm']
        aic_scores.append(gmm_model.aic(df[variables]))  # Usar las mismas columnas
        bic_scores.append(gmm_model.bic(df[variables]))  # Usar las mismas columnas

    # Graficar los resultados de AIC y BIC
    plt.figure(figsize=(10, 6))
    plt.plot(n_clusters_range, aic_scores, label='AIC', marker='o')
    plt.plot(n_clusters_range, bic_scores, label='BIC', marker='o')
    plt.title('Comparación de AIC y BIC para seleccionar el número de clústers')
    plt.xlabel('Número de clústers')
    plt.ylabel('Puntuación')
    plt.legend()
    plt.show()



In [ ]:
# Uso de la función
variables_usadas = ['Inflight wifi service', 'Inflight entertainment', 'Cleanliness', 'Departure Delay in Minutes']
n_clusters_range = range(3, 9)

evaluate_gmm_aic_bic(df_transformed, variables_usadas, n_clusters_range)

**Explique cuál sería el criterio adecuado para seleccionar el número óptimo de clústers. Justifique de forma estadística y a través de gráficos.**

El AIC (Criterio de Información de Akaike) y BIC (Criterio de Información Bayesiano) son criterios para seleccionar números de clusters. Ambos se utilizan para comparar modelos, castigando la complejidad del modelo (número de clusters en este caso) para evitar el sobre ajuste. El valor óptimo de clústers se asocia con el punto donde estas métricas son más bajas. Las métricas están hechas para comparar, por lo que más que el valor, lo que nos sirve es la diferencia entre las distintas opciones que generan AIC y BIC entre 3 y 8 clusters. En este caso, dado que las variables están regularizadas por el mínimo y máximo, dan ambas métricas en el rango de los negativos (a continuación se muestra el ejemplo, pero sin regularizar y el resultado de las métricas está en el rango de los positivos).

La interpretación no cambia: mientras menor sean los valores de los índices, es mejor el número de clusters, por lo que en este caso seria con 6 clusters.


In [ ]:
# Uso de la función con los datos sin regularizar
variables_usadas = ['Inflight wifi service', 'Inflight entertainment', 'Cleanliness', 'Departure Delay in Minutes']
n_clusters_range = range(3, 9)

evaluate_gmm_aic_bic(df_aerolineas_num, variables_usadas, n_clusters_range)

## 6. Análisis de resultados 📊 [10 puntos]

<center>
<img src="https://i.gifer.com/7wTk.gif" width=300>

Una vez identificado el número óptimo de clústers, se le pide realizar lo siguiente:

1. Utilizar la proyección en dos dimensiones para visualizar cada clúster claramente. [2 puntos]

2. ¿Es posible distinguir claramente entre los clústers generados? [2 puntos]

3. Proporcionar una descripción breve de cada clúster utilizando estadísticas descriptivas básicas, como la media y la desviación estándar, para resumir las características de las variables utilizadas en estos algoritmos. [2 puntos]

4. Proceda a visualizar los clústers en tres dimensiones para una perspectiva más detallada. [2 puntos]

5. ¿Cómo afecta esto a sus conclusiones anteriores? [2 puntos]

**Respuestas:**

In [160]:
def generate_clustered_df(df, variables, n_clusters):
    """
    Genera un DataFrame que incluye solo las variables seleccionadas, las etiquetas de clúster,
    y las primeras dos componentes principales (PC1 y PC2).
    
    Parámetros:
        df (DataFrame): DataFrame que contiene los datos.
        variables (list): Lista de nombres de las columnas a usar.
        n_clusters (int): Número de clústers elegidos.
    
    Retorna:
        DataFrame: DataFrame con las variables seleccionadas, los clústers, PC1 y PC2.
    """
    # Definir el pipeline
    pipeline = Pipeline([
        ('gmm', GaussianMixture(n_components=n_clusters, random_state=42))
    ])
    
    # Ajustar el modelo
    pipeline.fit(df[variables])
    
    # Predecir los clústers
    labels = pipeline.predict(df[variables])
    
    # Aplicar PCA para obtener las primeras dos componentes principales
    pca = PCA(n_components=2)
    principal_components = pca.fit_transform(df[variables])
    
    # Crear un DataFrame con las variables seleccionadas
    selected_vars_df = df[variables].copy()
    
    # Crear un DataFrame con los componentes principales
    pc_df = pd.DataFrame(principal_components, columns=['PC1', 'PC2'])
    
    # Agregar las etiquetas de clúster y las componentes principales al DataFrame
    selected_vars_df['Cluster'] = labels
    selected_vars_df = pd.concat([selected_vars_df, pc_df], axis=1)
    
    return selected_vars_df

In [161]:
# Uso de la función:
variables_usadas = ['Inflight wifi service', 'Inflight entertainment', 'Cleanliness', 'Departure Delay in Minutes']
n_clusters_elegido = 6  # Cambia este valor según lo que elijas

# Generar el DataFrame con clústeres y componentes principales
df_clustered = generate_clustered_df(df_transformed, variables_usadas, n_clusters_elegido)

# Mostrar el DataFrame resultante
print(df_clustered.head())

   Inflight wifi service  Inflight entertainment  Cleanliness  \
0                    0.6                     1.0          1.0   
1                    0.6                     0.2          0.2   
2                    0.4                     1.0          1.0   
3                    0.4                     0.4          0.4   
4                    0.6                     0.6          0.6   

   Departure Delay in Minutes  Cluster       PC1       PC2  
0                    0.015704        1 -0.467886  0.094105  
1                    0.000628        0  0.609520 -0.251048  
2                    0.000000        0 -0.407051  0.283905  
3                    0.006910        1  0.401034  0.025052  
4                    0.000000        0  0.070803 -0.078477  


In [162]:
def plot_pca_clusters(df_clustered):
    """
    Grafica las primeras dos componentes principales (PC1 y PC2) y colorea los puntos según el clúster.

    Parámetros:
        df_clustered (DataFrame): DataFrame que contiene las columnas 'PC1', 'PC2', y 'Cluster'.
    """
    # Graficar las componentes principales con el color basado en los clústers
    fig = px.scatter(
        df_clustered, x='PC1', y='PC2', color='Cluster', 
        title='Visualización de Clústers usando PCA',
        labels={'PC1': 'Componente Principal 1', 'PC2': 'Componente Principal 2'},
        color_continuous_scale=px.colors.qualitative.Set1  # Colores separados en lugar de un gradiente
    )
    
    # Mostrar el gráfico
    fig.show()

In [ ]:
# Uso de la función:
plot_pca_clusters(df_clustered)

**2.¿Es posible distinguir claramente entre los clústers generados?**

No es posible distinguir claramente, ya que no es tan claro en plano (X e Y) el sentido que tienen los clusters. Haría falta una dimensión más para verificar los diferentes clusters.

In [172]:
#3. Proporcionar una descripción breve de cada clúster utilizando estadísticas descriptivas básicas, 
#como la media y la desviación estándar, para resumir las características de las variables utilizadas en estos algoritmos

def estadistica(df_clustered, variables):
    """
    Calcula estadísticas descriptivas combinadas (media, desviación estándar, min, max, cuartiles)
    para cada clúster considerando todas las variables seleccionadas.

    Parámetros:
    df_clustered (DataFrame): DataFrame que contiene los datos con una columna 'Cluster'.
    variables (list): Lista de nombres de las columnas a usar para el análisis.
    """
    # Agrupar por clúster
    clusters = df_clustered.groupby('Cluster')

    # Crear una lista para almacenar los resultados
    results = []

    # Calcular estadísticas combinadas para cada clúster
    for cluster, data in clusters:
        # Subconjunto del dataframe con las variables seleccionadas
        subset = data[variables]

        # Verificar que no todos los valores sean NaN
        if not subset.isna().all().all():
            # Calcular las estadísticas combinadas
            mean_value = subset.mean().mean()  # Media combinada de todas las variables
            std_value = subset.stack().std()   # Desviación estándar combinada
            min_value = subset.min().min()     # Mínimo de todas las variables
            q25_value = subset.stack().quantile(0.25)  # Cuartil 25%
            q50_value = subset.stack().quantile(0.5)   # Mediana (cuartil 50%)
            q75_value = subset.stack().quantile(0.75)  # Cuartil 75%
            max_value = subset.max().max()     # Máximo de todas las variables

            # Agregar los resultados a la lista
            results.append({
                'Cluster': cluster,
                'Mean': mean_value,
                'Std': std_value,
                'Min': min_value,
                '25%': q25_value,
                '50%': q50_value,
                '75%': q75_value,
                'Max': max_value
            })

    # Convertir la lista de resultados a un DataFrame
    cluster_stats = pd.DataFrame(results)

    # Mostrar las estadísticas combinadas por clúster
    print(cluster_stats)

   Cluster      Mean       Std  Min       25%  50%  75%  Max
0        0  0.467032  0.357987  0.0  0.003141  0.4  0.8  1.0
1        1  0.464832  0.351388  0.0  0.075377  0.4  0.8  1.0
2        2  0.421208  0.338360  0.0  0.003141  0.4  0.8  1.0
3        3  0.470565  0.333355  0.0  0.020729  0.6  0.6  1.0
4        4  0.598721  0.399944  0.0  0.030151  0.8  1.0  1.0
5        5  0.452620  0.324643  0.0  0.195823  0.4  0.8  1.0


**Análisis de datos**

* Cluster 0: este clúster muestra un valor medio de 0.467, que es relativamente alto, pero tiene una desviación estándar considerable, lo que indica cierta dispersión en los datos. El 25% de los datos se encuentran muy cerca de 0, pero la mediana sugiere que la mayoría de los datos están por encima de 0.4.

* Cluster 1: Similar al clúster 0, el clúster 1 tiene una media y desviación estándar muy cercanas. La distribución parece tener una gran concentración en los valores más bajos, con un 25% de los datos apenas superando 0.075.

* Cluster 2: Este clúster tiene la media más baja entre los clústers analizados. Aunque la mediana es igual a la de los clústers anteriores, la media más baja sugiere que hay más valores extremos a la baja.

* Cluster 3: Este clúster muestra un valor medio alto, similar al clúster 0. La mediana de 0.6 indica que más de la mitad de los datos están en la parte superior de la escala. Esto podría indicar un perfil de clúster con mayores satisfacciones o condiciones.

* Cluster 4: Este clúster presenta la media más alta y la mayoría de los datos se encuentran en la parte superior del rango. Esto sugiere que este grupo tiene características muy favorables en las variables consideradas.

* Cluster 5: Este clúster tiene características intermedias, con una media que es bastante cercana a la de los clústers 0 y 1. La mediana indica que al menos el 50% de los datos están en la parte baja a media de la escala.


In [182]:
#4.Proceda a visualizar los clústers en tres dimensiones para una perspectiva más detallada.
import pandas as pd
from sklearn.decomposition import PCA

def generate_pca(df, variables, n_components=3):
    """
    Genera componentes principales utilizando PCA y agrega al DataFrame original.

    Parámetros:
        df (DataFrame): DataFrame que contiene los datos y los clústeres.
        variables (list): Lista de nombres de las columnas a usar.
        n_components (int): Número de componentes principales a generar.

    Retorna:
        DataFrame: DataFrame con las componentes principales y las etiquetas de los clústeres.
    """
    # Ajustar PCA
    pca = PCA(n_components=n_components)
    principal_components = pca.fit_transform(df[variables])
    
    # Crear un DataFrame para las componentes principales
    pca_df = pd.DataFrame(data=principal_components, columns=[f'PC{i+1}' for i in range(n_components)])
    
    # Agregar las etiquetas de clúster al nuevo DataFrame
    pca_df['Cluster'] = df['Cluster'].values
    
    return pca_df

def plot_3d_pca(pca_df):
    """
    Grafica las componentes principales en 3D, coloreadas por los clústeres.

    Parámetros:
        pca_df (DataFrame): DataFrame que contiene las componentes principales y las etiquetas de los clústeres.
    """
    # Crear una lista de colores separados
    unique_clusters = pca_df['Cluster'].unique()
    colors = px.colors.qualitative.Set1[:len(unique_clusters)]  # Selecciona colores de la paleta

    fig = px.scatter_3d(
        pca_df, 
        x='PC1', 
        y='PC2', 
        z='PC3', 
        color='Cluster',
        color_discrete_sequence=colors,  # Usar colores separados
        title='Componentes Principales en 3D',
        labels={'PC1': 'Componente Principal 1', 'PC2': 'Componente Principal 2', 'PC3': 'Componente Principal 3'}
    )
    fig.update_traces(marker=dict(size=5))
    fig.show()




In [ ]:
# Variables seleccionadas para PCA
variables_usadas = ['Inflight wifi service', 'Inflight entertainment', 'Cleanliness', 'Departure Delay in Minutes']

# Generar el DataFrame con las componentes principales
pca_df = generate_pca(df_clustered, variables_usadas)

# Graficar las componentes principales
plot_3d_pca(pca_df)


**5.¿Cómo afecta esto a sus conclusiones anteriores?**

Afecta Bastante las conclusiones anteriores, ya que en un plano 2D no se lograban entender del todo los clusters encontrados, pero al agregar una tercera dimensión y realizando una inspección gráfica a los datos, sí cobra sentido el agrupamiento que realizó y sus métricas. Se pueden ver los clusters más alejados, además del clúster principal.